### ISSUE TITLE
[Question] Is there any way to override the HttpMessage#CancellationToken?

### ISSUE THREAD

**Issue by dxynnez** [Azure/azure-sdk-for-net#23387]
**Question**
I am trying to add a custom timeout HttpPipelinePolicy by leveraging Polly's Policy.TimeoutAsync policy but found out that there is no way to override the Cancellation token in the HttpMessage nor have the delegate (ProcessNextAsync) to honor the cancellation token generated by the Timeout policy, hence there is no way to use the [Optimistic timeout strategy](https://github.com/App-vNext/Polly/wiki/Timeout#optimistic-timeout). Basically the problem is the same as https://stackoverflow.com/questions/55459962/how-to-enable-timeout-policy-with-polly 

Ideally if there is a way to reset the Cancellation token in the pipeline to the one that generated by the timeout policy then this should work.

Alternatively, is there any existing timeout HttpPipelinePolicy provided by azure sdk that I can easily plug into random ClientOptions then it would also work but I just couldn't find any.

**Environment:**
 - Azure.Core v1.17

**Comment by jsquire** [comment]
Thank you for your feedback.  Tagging and routing to the team member best able to assist.

**Comment by pakrym** [comment]
>Alternatively, is there any existing timeout HttpPipelinePolicy provided by azure sdk that I can easily plug into random ClientOptions then it would also work but I just couldn't find any.

There is existing support for timeouts. The https://docs.microsoft.com/en-us/dotnet/api/azure.core.retryoptions.networktimeout?view=azure-dotnet#Azure_Core_RetryOptions_NetworkTimeout is what used as a timeout for network operations. It's applied to request sending, response buffering or unbuffered response reads.

**Comment by dxynnez** [comment]
Thanks @jsquire and @pakrym ! Does this timeout apply to the first request (not the retry)? Also is there any way to also configure the PerCall timeout (per pipeline invocation)?

**Comment by pakrym** [comment]
>Does this timeout apply to the first request (not the retry)? 

It applies to every request individually.

>Also is there any way to also configure the PerCall timeout (per pipeline invocation)?

Yes:

``` C#
CancellationTokenSource cts = new();
cts.CancelAfter(...); // <---

client.CallAsync(..., cts.Token);
```

**Comment by dxynnez** [comment]
> > Does this timeout apply to the first request (not the retry)?
> 
> It applies to every request individually (first and retires).
> 
> > Also is there any way to also configure the PerCall timeout (per pipeline invocation)?
> 
> Yes but only by passing a CancellationToken to the client method:
> 
> ```cs
> CancellationTokenSource cts = new();
> cts.CancelAfter(...); // <---
> 
> client.CallAsync(..., cts.Token);
> ```
> 
> To answer your original question `Is there any way to override the HttpMessage#CancellationToken`, the token you pass to the client method call becomes the `HttpMessage#CancellationToken`

Thanks @pakrym ! I will put the client call into the Timeout policy callback then! This is a bit different than the original HttpMessageHandler pipeline which kinda defeat the purpose of HttpClientFactory and the ecosystem around it (e.g. Polly resiliency extension) - Is it recommended to use the new Azure SDK as singleton and have the SDK to manage the connection pooling instead? What would be the use case for customizing the HttpPipelineTransport nowadays? I see that it now default to a HttpClient delegate, which, seems to be only taking care of the socket connection.

**Comment by pakrym** [comment]
>Is it recommended to use the new Azure SDK as singleton and have the SDK to manage the connection pooling instead? 

Yes. Our goal is for an Azure SDK on its own to be efficient and reliable and solve all the common pitfalls people hit with HTTP client (pooling, limits, timeouts, connection recycling, etc.). 

>What would be the use case for customizing the HttpPipelineTransport nowadays? 

Not many. Advanced scenarios like setting a proxy, configuring custom limits, overriding cert validation.
